### ISSUE TITLE
Systemctl startup problems for rsyslog-8.31 on RHEL7

### ISSUE THREAD

**Issue by karloluiten** [rsyslog/rsyslog#2134]
Package: rsyslog-8.31.0-1.el7.x86_64
Kernel: 3.10.0-693.5.2.el7.x86_64
Strace: https://pastebin.com/raw/nLj70nvk

What I do:
systemctl start rsyslog

What I expect:
Startup within a few seconds

What happens:
Systemctl times out: "Job for rsyslog.service failed because a timeout was exceeded. See "systemctl status rsyslog.service" and "journalctl -xe" for details.".
Afterwards, the process does run and accepts syslog.

**Comment by karloluiten** [comment]
Output of rsyslogd -d: https://pastebin.com/raw/t8AKu7vw

**Comment by rgerhards** [comment]
many thanks for reporting this. I'll have a first look at it today and then we see if it's packaging-related or actual source code.

**Comment by rgerhards** [comment]
question: was that system rebooted after the install? I already have an idea based on the debug log.

Does ```journalctl -xe``` and ```systemctl status rsyslog``` show any error messages? (especially in regard to pid file?)

**Comment by rgerhards** [comment]
high probability that it is packaging related, just FYI

**Comment by karloluiten** [comment]
-xe:
```Nov 29 09:26:33 reporting.comp.local systemd[1]: Starting System Logging Service...
-- Subject: Unit rsyslog.service has begun start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit rsyslog.service has begun starting up.
Nov 29 09:26:33 reporting.comp.local audispd[707]: node=reporting.comp.local type=AVC msg=audit(1511943993.592:1364): avc:  denied  { unlink } for  pid=9410 comm="rsyslogd" name="log" dev="devtmpfs" ino=44254 scontext=system_u:system_r:syslogd_t:s0 tcontext=unconfined_u:object_r:device_t:s0 tclass=sock_file
Nov 29 09:26:33 reporting.comp.local audispd[707]: node=reporting.comp.local type=SYSCALL msg=audit(1511943993.592:1364): arch=c000003e syscall=87 success=yes exit=0 a0=7fbd55d5abef a1=1 a2=0 a3=2 items=2 ppid=1 pid=9410 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="rsyslogd" exe="/usr/sbin/rsyslogd" subj=system_u:system_r:syslogd_t:s0 key=(null)
Nov 29 09:26:33 reporting.comp.local audispd[707]: node=reporting.comp.local type=CWD msg=audit(1511943993.592:1364):  cwd="/"
Nov 29 09:26:33 reporting.comp.local audispd[707]: node=reporting.comp.local type=PATH msg=audit(1511943993.592:1364): item=0 name="/dev/" inode=3 dev=00:05 mode=040755 ouid=0 ogid=0 rdev=00:00 obj=system_u:object_r:device_t:s0 objtype=PARENT
Nov 29 09:26:33 reporting.comp.local audispd[707]: node=reporting.comp.local type=PATH msg=audit(1511943993.592:1364): item=1 name="/dev/log" inode=44254 dev=00:05 mode=0140666 ouid=0 ogid=0 rdev=00:00 obj=unconfined_u:object_r:device_t:s0 objtype=DELETE
Nov 29 09:26:33 reporting.comp.local audispd[707]: node=reporting.comp.local type=PROCTITLE msg=audit(1511943993.592:1364): proctitle=2F7362696E2F727379736C6F6764002D6E
Nov 29 09:26:33 reporting.comp.local audispd[707]: node=reporting.comp.local type=EOE msg=audit(1511943993.592:1364):
Nov 29 09:26:37 reporting.comp.local audispd[707]: node=reporting.comp.local type=PROCTITLE msg=audit(1511943997.165:1365): proctitle=2F7573722F62696E2F707974686F6E002F62696E2F6175746F6A756D70002D2D616464002F726F6F74
Nov 29 09:26:37 reporting.comp.local audispd[707]: node=reporting.comp.local type=EOE msg=audit(1511943997.165:1365):
Nov 29 09:26:45 reporting.comp.local audispd[707]: node=reporting.comp.local type=PROCTITLE msg=audit(1511944005.060:1366): proctitle=2F7573722F62696E2F707974686F6E002F62696E2F6175746F6A756D70002D2D616464002F726F6F74
Nov 29 09:26:45 reporting.comp.local audispd[707]: node=reporting.comp.local type=EOE msg=audit(1511944005.060:1366):```

status:
```root@reporting:~# systemctl status rsyslog
● rsyslog.service - System Logging Service
   Loaded: loaded (/usr/lib/systemd/system/rsyslog.service; enabled; vendor preset: enabled)
   Active: activating (start) since Wed 2017-11-29 09:28:03 CET; 30s ago
     Docs: man:rsyslogd(8)
           http://www.rsyslog.com/doc/
 Main PID: 10007 (rsyslogd)
   CGroup: /system.slice/rsyslog.service
           └─10007 /sbin/rsyslogd -n

Nov 29 09:28:03 reporting.kahuna.local systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 09:28:03 reporting.kahuna.local systemd[1]: Starting System Logging Service...
```

**Comment by karloluiten** [comment]
System was rebooted indeed.

**Comment by rgerhards** [comment]
> System was rebooted indeed.

ok - actually I had originally thought that a reboot would resolve a temporary glitch in the package. But probably the systemd unit file is not maintained properly (still only a guess).

**Comment by rgerhards** [comment]
oops.. that is auditd ;-)

**Comment by karloluiten** [comment]
Deleted ;)

```
[Unit]
Description=System Logging Service
;Requires=syslog.socket
Documentation=man:rsyslogd(8)
Documentation=http://www.rsyslog.com/doc/

[Service]
Type=notify
ExecStart=/sbin/rsyslogd -n
StandardOutput=null
Restart=on-failure
UMask=0066

[Install]
WantedBy=multi-user.target
;Alias=syslog.service
```
Does not look all to spectacular...

**Comment by rgerhards** [comment]
Indeed... I know that @friedl applied some changes re an invalid pid file path. Thought that may be related, but doesn't sound like it is used. So I need to spawn a real lab.

If you grep through the journal, do you see any of these messages (they are from the debug log):
```
rsyslogd: pidfile '/var/run/syslogd.pid' and pid 7568 already exist.
If you want to run multiple instances of rsyslog, you need to specify
different pid files for them (-i option).
rsyslogd: run failed with error -3000 (see rsyslog.h or try http://www.rsyslog.com/e/3000 to learn what that number means)
```
That's a good pointer to the cause, and I had expected that they show up pretty early in journal...

**Comment by karloluiten** [comment]
```
[klu@reporting ~]$ sudo journalctl | grep rsys
Nov 29 10:09:01 reporting.comp.local audispd[704]: node=reporting.comp.local type=SYSCALL msg=audit(1511946541.269:49): arch=c000003e syscall=2 success=yes exit=14 a0=7f01100147f0 a1=80541 a2=1a0 a3=2 items=2 ppid=1 pid=731 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm=72733A6D61696E20513A526567 exe="/usr/sbin/rsyslogd" subj=system_u:system_r:syslogd_t:s0 key="logins"
Nov 29 10:10:31 reporting.comp.local systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:10:31 reporting.comp.local systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:10:31 reporting.comp.local systemd[1]: rsyslog.service failed.
Nov 29 10:10:31 reporting.comp.local audispd[704]: node=reporting.comp.local type=SERVICE_START msg=audit(1511946631.121:150): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=rsyslog comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
Nov 29 10:10:31 reporting.comp.local systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:10:31 reporting.comp.local audispd[704]: node=reporting.comp.local type=SERVICE_START msg=audit(1511946631.301:156): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=rsyslog comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Nov 29 10:10:31 reporting.comp.local audispd[704]: node=reporting.comp.local type=SERVICE_STOP msg=audit(1511946631.301:157): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=rsyslog comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Nov 29 10:10:37 reporting.comp.local audispd[704]: node=reporting.comp.local type=SYSCALL msg=audit(1511946637.873:166): arch=c000003e syscall=2 success=yes exit=14 a0=7f798c00a9f0 a1=80541 a2=1a0 a3=2 items=2 ppid=1 pid=1845 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm=72733A6D61696E20513A526567 exe="/usr/sbin/rsyslogd" subj=system_u:system_r:syslogd_t:s0 key="logins"
[klu@reporting ~]$ sudo journalctl | grep pidfile
[klu@reporting ~]$ 
```

Output after a reboot. Here I do see some auditd-problems. Odd.

**Comment by rgerhards** [comment]
It looks like we have an issue reporting the startup failure messages to journald properly. My gut feeling is that rsyslog is the culprit. So actually a new issue, but I'll keep it on this tracker, at least until it is confirmed.

**Comment by ghubusr01** [comment]
I have similiar problems on CentOS 7.4 here since tonight. It is on a production server

[root@ms1 log]# journalctl | grep rsys
Nov 29 10:20:58 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:20:58 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:20:58 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:20:58 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:22:29 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:22:29 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:22:29 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:22:29 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:23:59 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:23:59 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:23:59 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:23:59 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:25:30 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:25:30 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:25:30 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:25:30 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:27:00 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:27:00 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:27:00 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:27:00 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:28:30 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:28:30 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:28:30 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:28:31 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:30:01 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:30:01 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:30:01 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:30:01 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:31:31 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:31:31 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:31:31 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:31:31 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:33:02 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:33:02 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:33:02 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:33:02 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:34:32 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:34:32 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:34:32 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:34:32 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.
Nov 29 10:36:03 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:36:03 ms1.p.tomgugel.com systemd[1]: Unit rsyslog.service entered failed state.
Nov 29 10:36:03 ms1.p.tomgugel.com systemd[1]: rsyslog.service failed.
Nov 29 10:36:03 ms1.p.tomgugel.com systemd[1]: rsyslog.service holdoff time over, scheduling restart.

This is from journalctl -xe
Nov 29 10:40:34 ms1.p.tomgugel.com systemd[1]: rsyslog.service start operation timed out. Terminating.
Nov 29 10:40:34 ms1.p.tomgugel.com systemd[1]: Failed to start System Logging Service.

The server was also rebooted a few minutes ago that did not solve a problem. Rsyslog config files weren't changed also

**Comment by ghubusr01** [comment]
Actually it's the same on all my production machines :(

**Comment by rgerhards** [comment]
I can't get you a fix immediately. can you nail 8.30? in the mean time.

**Comment by ghubusr01** [comment]
A yum downgrade did it for the moment

[root@ms1 system]# systemctl status rsyslog
● rsyslog.service - System Logging Service
   Loaded: loaded (/usr/lib/systemd/system/rsyslog.service; enabled; vendor preset: enabled)
   **Active: active (running) since Wed 2017-11-29 11:24:10 CET; 5s ago**
     Docs: man:rsyslogd(8)
           http://www.rsyslog.com/doc/
 Main PID: 17719 (rsyslogd)
   CGroup: /system.slice/rsyslog.service
           └─17719 /sbin/rsyslogd -n

**Comment by ghubusr01** [comment]
I excluded rsyslog in yum.conf for the moment until this is fixed, please provide according information.

Thank you!

**Comment by Dolbager** [comment]
Same issue, to fix need to downgrade and lock version . 
OS: Centos 7.4
Kernel: 4.14.2-1.el7.elrepo.x86_64
Systemd: 219-42.el7_4.4

```
yum install yum-plugin-versionlock -y
yum remove rsyslog -y
yum install rsyslog-8.30.0-4.el7 -y
yum versionlock rsyslog
```

**Comment by friedl** [comment]
Rainer and I went back and forth on this bug. But we were successful in creating a good workaround. This has been tested and rsyslog starts fine in my lab.

I have already built and uploaded the new RPMs, which should now identify as rsyslog-8.30.0-2.

**Comment by ghubusr01** [comment]
You mean 8.31.0-2 right? It seems to work here on CentOS 7.4

**Comment by friedl** [comment]
Sorry. 8.31.0-2 of course.

Edited the comment.

**Comment by ghubusr01** [comment]
Thanks for the quick fix

**Comment by rgerhards** [comment]
We did some further testing, and there is still one glitch with ```systemctl start rsyslog``` while ```systemctl restart rsyslog``` works. We are looking into this one as well.

**Comment by ghubusr01** [comment]
ok so that means as long as the server isn't rebooted the actual version can stay in production?

**Comment by rgerhards** [comment]
right now running tests...

**Comment by rgerhards** [comment]
sorry, yes, seems to also hinder startup :-(

**Comment by rgerhards** [comment]
to be a bit more precise, I (under vagrant) see issues whereas @friedl  (und VMWare) does not have any. Anyhow, I hope to have a fix underway as I write this...

**Comment by friedl** [comment]
I uploaded a new version 8.31.0-3 just now with a potential fix by Rainer for the new service start issue.
But, I cannot confirm if `systemctl start rsyslog` works properly now, because I could not see it before.

**Comment by ghubusr01** [comment]
There is no 8.31.0-3 version in the repo

**Comment by deoren** [comment]
Not sure if it's helpful feedback, but when @friedl  first announced that the `8.31.0-3` version was available I tested then on a stock CentOS 7 VM. The package was available for installation and installed seemingly without issue. 

At that point I was able to stop, start and restart the daemon without any warnings/errors in `/var/log/messages` or the output of `systemctl status rsyslog`.

**Comment by ghubusr01** [comment]
I tested it on a KVM and an OpenVZ install in production, both with CentOS 7.4 (latest) and latest CPanel version. There it is definitely not working, resulting with the above errors.

**Comment by specnazzz** [comment]
@ghubusr01 There was a change in the `rsyslog.service` file. I think you need to run `systemctl disable rsyslog` and then `systemctl enable rsyslog` so that the `syslog.service` symlink is properly created. Then start/restart should work.

**Comment by ghubusr01** [comment]
Thanks for that hint @specnazzz this did it!

**Comment by karloluiten** [comment]
The disable/enable "trick" seems to work, though this is merely a workaround, I would not like to do this to all my servers separately.

**Comment by deoren** [comment]
As others have said, I too can report success with disabling and then re-enabling the rsyslog.service unit on a CentOS 7 system.

`# rpm -q rsyslog`
```
rsyslog-8.31.0-3.el7.x86_64
```

**Comment by rgerhards** [comment]
Just let me add that it looks like the workaround provided does do the job. It's not a bad workaround, but we are still trying to solve the root cause in the package build process. For some reason, the build seems not to install the (required) systemd-devel package, which then leads to code that does not properly support the advanced communication mode between rsyslog and systemd that was configured inside the service file.

If someone could lend us a helping hand in solving this, we would be all ears ;-)

**Comment by rgerhards** [comment]
side-note: ```systemctl daemon-reload``` should also do the trick.

**Comment by deoren** [comment]
> @rgerhards: side-note: systemctl daemon-reload should also do the trick (for the work-around).

Unfortunately I just tested and found that it wasn't sufficient to solve the problem. 

1. Boot CentOS 7 VM with stock rsyslog from CentOS repos
1. Install rsyslog via use of http://rpms.adiscon.com/v8-stable/rsyslog.repo
1. Reboot
1. Confirm that rsyslog did not start
1. `systemctl daemon-reload`
1. Reboot
1. Confirm that rsyslog did not start

**Comment by rgerhards** [comment]
oh, interesting to hear. Always worked for me in those cases.

**Comment by deoren** [comment]
@rgerhards 

From what I can tell, running `systemctl daemon-reload` is primarily useful for when the on-disk contents of the primary or drop-in conf files change and you need to apply the changes right away, without a reboot or (here is where my knowledge ends) some other scheduled process forces systemd to re-read the configuration.

Take my words with a grain of salt as I am still new to working with systemd. ;)

**Comment by deoren** [comment]
Not sure if it's helpful, but when disabling/renabling the unit this stood out to me.

## First time disabling/re-enabling rsyslog.service unit

`[root@localhost redhat]# systemctl disable rsyslog`
```
Removed symlink /etc/systemd/system/multi-user.target.wants/rsyslog.service.
```

`[root@localhost redhat]# systemctl enable rsyslog`
```
Created symlink from /etc/systemd/system/syslog.service to /usr/lib/systemd/system/rsyslog.service.
Created symlink from /etc/systemd/system/multi-user.target.wants/rsyslog.service to /usr/lib/systemd/system/rsyslog.service.
```

## Second time disabling/re-enabling rsyslog.service unit

`[root@localhost yum.repos.d]# systemctl disable rsyslog`
```
Removed symlink /etc/systemd/system/multi-user.target.wants/rsyslog.service.
Removed symlink /etc/systemd/system/syslog.service.
```

`[root@localhost yum.repos.d]# systemctl enable rsyslog`
```
Created symlink from /etc/systemd/system/syslog.service to /usr/lib/systemd/system/rsyslog.service.
Created symlink from /etc/systemd/system/multi-user.target.wants/rsyslog.service to /usr/lib/systemd/system/rsyslog.service.
```

The `/etc/systemd/system/syslog.service to /usr/lib/systemd/system/rsyslog.service` symbolic link is not created when installing the latest package, but is created when disabling/re-enabling the unit.

Is this relevant?

I notice that the CentOS 7 upstream rsyslog.service file has this line commented out:

```
[Install]
WantedBy=multi.user.target
;Alias=syslog.service
```

whereas the Adiscon repo package does not:

```
[Install]
WantedBy=multi.user.target
Alias=syslog.service
```

**Comment by deoren** [comment]
From https://git.centos.org/blob/rpms!rsyslog.git/c7/SPECS!rsyslog.spec:

```
%systemd_post rsyslog.service

%preun
%systemd_preun rsyslog.service

%postun
%systemd_postun_with_restart rsyslog.service
```

I looked further and found that the macros are defined like so in the [`macros.systemd.in`](https://github.com/systemd/systemd/blob/fc11a1a2041a9050e4aab7fa7bde67f5fd6a3a73/src/core/macros.systemd.in#L51) file:



```
%systemd_post() \
if [ $1 -eq 1 ] ; then \
        # Initial installation \
        systemctl --no-reload preset %{?*} >/dev/null 2>&1 || : \
fi \
%{nil}
```

Running that command manually seems to provide the missing symbolic link:

`[root@localhost redhat]# systemctl --no-reload preset rsyslog`
```
Created symlink from /etc/systemd/system/syslog.service to /usr/lib/systemd/system/rsyslog.service.
```

**Comment by rgerhards** [comment]
@deoren (and everyone else ;)) it would be great if you could give a try the version @friedl posted here:
https://github.com/rsyslog/rsyslog/issues/2167#issuecomment-349934562

It possibly is the final fix for this issue here.

**Comment by deoren** [comment]
> @rgerhards: @deoren (and everyone else ;)) it would be great if you could give a try the version @friedl posted here:
> #2167 (comment)
> 
> It possibly is the final fix for this issue here.

Short version: **Same failures as before**

From a freshly updated CentOS 7 VM with CentOS-provided rsyslog (`rsyslog-8.24.0-12.el7`), I followed the steps provided by @friedl to enable access to the test package (`rsyslog-8.31.0-5.el7`) and install it.

## Configuration

I've made no adjustments. Posting for reference (if needed).

`[root@centos7-test yum.repos.d]# grep -Ev '^$|^\s*#' /etc/rsyslog.conf /etc/rsyslog.d/*.conf`

```
/etc/rsyslog.conf:module(load="imuxsock") # provides support for local system logging (e.g. via logger command)
/etc/rsyslog.conf:module(load="imklog")   # provides kernel logging support (previously done by rklogd)
/etc/rsyslog.conf:$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
/etc/rsyslog.conf:$IncludeConfig /etc/rsyslog.d/*.conf
/etc/rsyslog.conf:*.info;mail.none;authpriv.none;cron.none                /var/log/messages
/etc/rsyslog.conf:authpriv.*                                              /var/log/secure
/etc/rsyslog.conf:mail.*                                                  /var/log/maillog
/etc/rsyslog.conf:cron.*                                                  /var/log/cron
/etc/rsyslog.conf:*.emerg                                                 :omusrmsg:*
/etc/rsyslog.conf:uucp,news.crit                                          /var/log/spooler
/etc/rsyslog.conf:local7.*                                                /var/log/boot.log
```

```
/etc/rsyslog.d/listen.conf:$SystemLogSocketName /run/systemd/journal/syslog
```

## Immediately after installing `rsyslog-8.31.0-5.el7`

After installing via `yum install rsyslog-8.31.0`, here are the commands I ran and the associated output:

`[root@centos7-test yum.repos.d]# rpm -q rsyslog`
```
rsyslog-8.31.0-5.el7.centos.x86_64
```

`[root@centos7-test yum.repos.d]# systemctl status rsyslog`
```
● rsyslog.service - System Logging Service
   Loaded: loaded (/usr/lib/systemd/system/rsyslog.service; enabled; vendor preset: enabled)
   Active: active (running) since Thu 2017-12-07 08:28:44 CST; 2min 43s ago
     Docs: man:rsyslogd(8)
           http://www.rsyslog.com/doc/
 Main PID: 2453 (rsyslogd)
   CGroup: /system.slice/rsyslog.service
           └─2453 /sbin/rsyslogd -n

Dec 07 08:28:43 centos7-test systemd[1]: Starting System Logging Service...
Dec 07 08:28:44 centos7-test systemd[1]: Started System Logging Service.
```

`[root@centos7-test yum.repos.d]# systemctl status syslog.socket`
```
● syslog.socket - Syslog Socket
   Loaded: loaded (/usr/lib/systemd/system/syslog.socket; static; vendor preset: disabled)
   Active: inactive (dead)
     Docs: man:systemd.special(7)
           http://www.freedesktop.org/wiki/Software/systemd/syslog
   Listen: /run/systemd/journal/syslog (Datagram)

Dec 07 08:28:43 centos7-test systemd[1]: Socket service syslog.service not loaded, refusing.
Dec 07 08:28:43 centos7-test systemd[1]: Failed to listen on Syslog Socket.
```

`[root@centos7-test yum.repos.d]# ls -l /run/systemd/journal/syslog`
```
ls: cannot access /run/systemd/journal/syslog: No such file or directory
```

`[root@centos7-test yum.repos.d]# fuser -v /dev/log`
```
                     USER        PID ACCESS COMMAND
/dev/log:            root          1 F.... systemd
                     root        471 F.... systemd-journal
                     root       2453 F.... rsyslogd
```

If I'm reading that correctly then the rsyslogd process is binding (read/write?) to the `/dev/log` socket and the `/run/systemd/journal/syslog` socket is not available.


## Post-install workarounds

`[root@centos7-test yum.repos.d]#  systemctl --no-reload preset rsyslog`
```
Created symlink from /etc/systemd/system/syslog.service to /usr/lib/systemd/system/rsyslog.service.
```

I then rebooted the test VM. Attempts to restart either `syslog.socket` or `rsyslog.service` before rebooting all fail.

## Post-reboot details

`[root@centos7-test redhat]# systemctl status rsyslog`
```
● rsyslog.service - System Logging Service
   Loaded: loaded (/usr/lib/systemd/system/rsyslog.service; enabled; vendor preset: enabled)
   Active: active (running) since Thu 2017-12-07 11:02:19 CST; 2min 28s ago
     Docs: man:rsyslogd(8)
           http://www.rsyslog.com/doc/
 Main PID: 702 (rsyslogd)
   CGroup: /system.slice/rsyslog.service
           └─702 /sbin/rsyslogd -n

Dec 07 11:02:19 centos7-test systemd[1]: Starting System Logging Service...
Dec 07 11:02:19 centos7-test rsyslogd[702]: command 'SystemLogSocketName' is currently not permitted - did you already set it via a RainerScript command (v6+ config)? [v8.31.0 try http://www.rsyslog.com/e/2222 ]
Dec 07 11:02:19 centos7-test rsyslogd[702]:  [origin software="rsyslogd" swVersion="8.31.0" x-pid="702" x-info="http://www.rsyslog.com"] start
Dec 07 11:02:19 centos7-test systemd[1]: Started System Logging Service.
Hint: Some lines were ellipsized, use -l to show in full.
```

`[root@centos7-test redhat]# systemctl status syslog.socket`
```
● syslog.socket - Syslog Socket
   Loaded: loaded (/usr/lib/systemd/system/syslog.socket; static; vendor preset: disabled)
   Active: active (running) since Thu 2017-12-07 11:02:17 CST; 2min 35s ago
     Docs: man:systemd.special(7)
           http://www.freedesktop.org/wiki/Software/systemd/syslog
   Listen: /run/systemd/journal/syslog (Datagram)
```

`[root@centos7-test redhat]# ls -l /run/systemd/journal/syslog`
```
srw-rw-rw-. 1 root root 0 Dec  7 11:02 /run/systemd/journal/syslog
```

**Comment by rgerhards** [comment]
I currently consider this solved (based on feedback and no new reports). If it persists, let us know and I'll re-open.

**Comment by deoren** [comment]
@rgerhards The problem does not appear to be solved, though a workaround _is_ possible via one of these two methods:

1. `systemctl daemon-reload; systemctl disable rsyslog; systemctl enable rsyslog`
1. `systemctl --no-reload preset rsyslog`

I just tested and replicated what I reported previously: **If you take a fresh CentOS 7 system and install rsyslog from the rpms.adiscon.com repo, rsyslog will fail to start on boot**.

This is with the latest stable (`rsyslog-8.31.0-4.el7.centos.x86_64`) or testing (`rsyslog-8.31.0-5.el7.centos.x86_64`) versions of the package.

- [rsyslog_adiscon_repo_testing_install_results_2017-12-18.txt](https://github.com/rsyslog/rsyslog/files/1568192/rsyslog_adiscon_repo_testing_install_results_2017-12-18.txt)
- [rsyslog_adiscon_repo_stable_install_results_2017-12-18.txt](https://github.com/rsyslog/rsyslog/files/1568193/rsyslog_adiscon_repo_stable_install_results_2017-12-18.txt)

I've attached the commands run and the output for both the stable and testing versions of the package. If you'd rather have them posted in-line I can do that too.

**Comment by rgerhards** [comment]
@friedl can you please work with @deoren on the resolution, best before the holidays?

**Comment by deoren** [comment]
> @rgerhards:  @friedl can you please work with @deoren on the resolution, best before the holidays?

I'll be happy to test any further builds, just let me know. Thus far my test case has been upgrading from the stock CentOS 7 rsyslog package and from the earlier 8.31.0 stable package, but I'm also fine with testing any other combination that you would find helpful.

**Comment by rgerhards** [comment]
I just had a quick chat with @friedl - we need to work on @friedl's environment so that he can reproduce the issue. Right now, we don't see it (I am *not* saying it doesn't exist, just that *we* don't see it).

**Comment by deoren** [comment]
@rgerhards: I just had a quick chat with @friedl - we need to work on @friedl's environment so that he can reproduce the issue. Right now, we don't see it (I am not saying it doesn't exist, just that we don't see it).

I'll create a fresh CentOS 7 VM from scratch to make sure my environment doesn't have any oddities about it. I'll report back once I've done so.

**Comment by deoren** [comment]
Here are the kickstarter files for the new VM (.txt extension to comply with GitHub requirement):

- [anaconda-ks.cfg.txt](https://github.com/rsyslog/rsyslog/files/1568672/anaconda-ks.cfg.txt)
- [initial-setup-ks.cfg.txt](https://github.com/rsyslog/rsyslog/files/1568673/initial-setup-ks.cfg.txt)

I tested on a fresh VM (after applying all CentOS 7 updates) and the results are the same for the v8-stable and testing builds.

**Comment by friedl** [comment]
Thanks for the information. I'll see if I can rebuild accordingly, so we can probably see the same issue.

**Comment by deoren** [comment]
@friedl If you have any problems, let me know and I can upload the whole VM somewhere for you to grab. It's around 5.5 GB on my end.

**Comment by friedl** [comment]
@deoren good news. Though it took a while now, I was able to reproduce what you observed.

We also found some difference in the service file. Here is the original one:

```
[Unit]
Description=System Logging Service
;Requires=syslog.socket
Documentation=man:rsyslogd(8)
Documentation=http://www.rsyslog.com/doc/

[Service]
Type=notify
EnvironmentFile=-/etc/sysconfig/rsyslog
ExecStart=/usr/sbin/rsyslogd -n $SYSLOGD_OPTIONS
Restart=on-failure
UMask=0066
StandardOutput=null
Restart=on-failure

[Install]
WantedBy=multi-user.target
;Alias=syslog.service
```

And here is the one provided by our RPM:

```
[Unit]
Description=System Logging Service
Requires=syslog.socket
Documentation=man:rsyslogd(8)
Documentation=http://www.rsyslog.com/doc/

[Service]
Type=notify
EnvironmentFile=-/etc/sysconfig/rsyslog
ExecStart=/sbin/rsyslogd -n $SYSLOGD_OPTIONS
Restart=on-failure
UMask=0066
StandardOutput=null
Restart=on-failure

[Install]
WantedBy=multi-user.target
Alias=syslog.service
```

We'll look at this after the Christmas holidays. 

Perhaps it will be sufficient to just change the path for the executable. But, most likely, this will break everything, so we will do some test builds first.

**Comment by deoren** [comment]
> @friedl: @deoren good news. Though it took a while now, I was able to reproduce what you observed.

Nice!

> @friedl: We also found some difference in the service file.

I noticed that your unit file had the previously commented out `syslog.{socket, service}` items enabled (compared to the prior version), but I completely missed the difference in the path between the two. Oh well.

> @friedl: We'll look at this after the Christmas holidays.
> 
> Perhaps it will be sufficient to just change the path for the executable. But, most likely, this will break everything, so we will do some test builds first.

If/when you're ready to have me test a new package just let me know.

**Comment by friedl** [comment]
@deoren We tried using the original spec file from the official CentOS release. I tested it here and it looks good to me. 

Can you try to install rsyslog-8.31.0-6 from the Testing repository for confirmation, please?

If you could do that today, that would be great. We want to test some other things as well before the next release, so we'd like to have this resolved asap.

**Comment by deoren** [comment]
@friedl I will give it a try later today and post back.

**Comment by deoren** [comment]
## What I did

1. `su`
1. `cd /etc/yum.repos.d`
1. `wget http://rpms.adiscon.com/testing/rsyslog.repo`
1. `yum install rsyslog-8.31.0`

## Results

After rebooting I ran `systemctl status rsyslog` and saw that it failed to start. I ran `rsyslogd -N6` and got this back:

```
rsyslogd: symbol lookup error: rsyslogd: undefined symbol: fjson_global_do_case_sensitive_comparison
```

I confirmed that I had `rsyslog-8.31.0-6.el7.centos` installed.

Perhaps the new spec file does not apply a dependency on a newer version of libfastjson?

## Follow-up steps

I checked and found that `libfastjson-0.99.4-2.el7.x86_64` was installed. I ran `yum upgrade` and was offered the 8.32 rsyslog package, so I bailed and targeted just libfastjson with `yum install libfastjson`. That completed and rsyslog appeared to start.

I reverted to a snapshot where I had just dropped in the yum conf fragment, but had yet to call on `yum` to update its cache or install on any packages from the adiscon repo. I ran `yum install rsyslog-8.31.0 libfastjson` and then followed with `systemctl status rsyslog`. The rsyslog daemon was running at this point. Seems to be a dependency issue on a specific version of libfastjson as noted.

I rebooted and after the system came back up I ran `systemctl status rsyslog` again. The results appeared to be as expected.

## Observations

I noticed some changes between the testing and stable builds:

- The testing package (`rsyslog-8.31.0-6`)
    - relies on `imjournal`
    - uses the legacy format for the configuration
    - has `Unit` > `Requires=syslog.socket` commented out
    - has `Install` > `Alias=syslog.service` commented out
    - `/run/systemd/journal/syslog` is not created as a result of commented out items

- The current stable package  (`rsyslog-8.31.0-4`)
    - does not load `imjournal`
    - uses the modern/advanced format (at least for the load statements)
    - has `Unit` > `Requires=syslog.socket` uncommented
    - has `Install` > `Alias=syslog.service` uncommented
    - `/run/systemd/journal/syslog` is created when using one of the earlier workarounds (which allows imuxsock to be used in place of imjournal)

Are those changes intentional? If so, it sounds like anyone relying on the earlier package behavior of using imuxsock over imjournal to pull in messages from the system log socket will need to create unit file drop-ins to restore the `/run/systemd/journal/syslog` socket. 

Not the end of the world, but something worth noting, perhaps in the changelog (package or project level, not sure which).

**Comment by rgerhards** [comment]
> Are those changes intentional? If so, it sounds like anyone relying on the earlier package behavior of using imuxsock over imjournal to pull in messages from the system log socket will need to create unit file drop-ins to restore the /run/systemd/journal/syslog socket.

Yes, kind of. I requested these. Reason:

As far as I could track down the issue here, it had to do with our packages not being compatible to the CentOS provided ones. Trying to dig into the dark history of our packaging, it looks like we began to provide the CentOS 7 packages around the time RHEL 7 beta appeared. As we do not have access the RHEL, we seem to have continued to use the CentOS 6 spec for building these packages. On the other hand, RHEL did considerate changes, which we seen to never have adopted even after CentOS 7 became final (and so we got access to the actual spec file).

As far as I could diagnose, this lead to quite some incompatibilities between the CentOS 7 and our packages. I admit I wonder they ever worked and looking at this thread I am not sure if they really did.

I am tired of all these packaging issues. So I asked @friedl to pull in the original CentOS 7 spec files and build our packages based on them. As it looks, this seems to have solved the root problem, albeit the dependency for libfastjson is wrong [and I wonder why our test worked... again time to question the testing procedure :-( ].

We now have some collateral damage, e.g. the changes to new style that we made. This can be fixed later on. My understanding is that most people will not notice anyhow, as I think (but do not know) we do not overwrite the rsyslog.conf when our package is installed. Even if it is, I think it is worth it.

If I am not mistaken, this change should provide the easy and seamingless upgrade capability for CentOS 7 users that we had in mind -- just use a new rsyslog version, but do not touch the system configuration you previously had.

That said, I would still like to do some improvements, but I first would like to see that at least for one release the packaging works as well. It's really more than tiring to work hard on software quality and then have the packaging process thrash all the effort... ;-)

Feedback, of course, is happily accepted. Don't let you be turned off by my frustration. And - many thanks for the detailled response, it is most useful. We should document this with the release.

**Comment by ghubusr01** [comment]
Hmm, I just read this thoroughly but I am not sure what to to do now.
My configuration is actually running on CentOS 7.4 with 
module(load="imuxsock" SysSock.Name="/run/systemd/journal/syslog")
module(load="imklog")

Can I upgrade then now or do I have to change to imjournal something. That's not exactly clear for me and I would be happy for an answer.

Thanks.

**Comment by rgerhards** [comment]
@ghubusr01 did you modify rsyslog.conf yourself? My (possibly wrong) understanding is that the config is left untouched if you modified it and only overwritten if it is left default. So if you modified it, nothing should change.

unrelated side-note: we are looking for several years now for folk who would like to help with the packaging, as our best effort for sure is just that - best effort from folks who do not know precisely what they are doing. I was more often then not tempted to stop providing packages...

**Comment by ghubusr01** [comment]
@rgerhards Yes, when installing rsyslog from your repo at the beginning of 2017 I edited it myself after reading the docs and deciding how I wanted to run it.
Never changed it since that and got problems first time with the 8.31 version.

**Comment by friedl** [comment]
@deoren Thank you for your feedback. As Rainer said, the changes were intended. 

I updated the RPM a few more times today and fixed the dependency issue for libfastjson by forcing the requirement. In my lab with a fresh install, this seems to work fine now. Can you please confirm?

**Comment by rgerhards** [comment]
@ghubusr01 
> @rgerhards Yes, when installing rsyslog from your repo at the beginning of 2017 I edited it myself after reading the docs and deciding how I wanted to run it.

As of *my* understanding, you should be safe. But I recommend to ask someone with solid knowledge of the way rpms operate.

**Comment by ghubusr01** [comment]
@rgerhards Ok thanks so now I understand what you meant with this sentence with "...precisely...". I'll just give it a try, the question was less if my .conf file is overwritten (I can save that before) rather than what deoren said regarding creating unit file drop-ins for syslog socket.
@deoren Can you make this more clear for me please? Thanks

**Comment by rgerhards** [comment]
@ghubusr01 sorry, I overlooked the question on "...precisely..." -- it was probably my English! Glad it's solved.

**Comment by rgerhards** [comment]
@ghubusr01 and I think I understand your question better now: The old package modified system configuration, as we enabled socket activate in the systemd unit configuration, which CentOS 7 standard does not use. Again assuming that you made all your configuration changes, I would think that neither rsyslog.conf nor rsyslog.service is overwritten by the package.

**Comment by friedl** [comment]
@ghubusr01 I just tried and edited the default configuration before updating to a recent rsyslog version on a fresh system. My configuration file was not overwritten.

**Comment by ghubusr01** [comment]
@rgerhards  Ok thanks, I did not change systemd unit configuration rsyslog.service only rsyslog.conf. As said I will have a look on both files. I have only production servers at the moment due to costs.

What else I saw now is that in /etc/rsyslog.d/listen.conf the line $SystemLogSocketName /run/systemd/journal/syslog was in there but it was copied to listen.conf.save on the 29th (where I installed 8.31 and then reverted back to 8.30) and listen.conf is now empty.

**Comment by deoren** [comment]
@ghubusr01,

Not able to test at the moment, but I hope to do so tonight or some point Saturday or Sunday.

As I understand it, when the new version drops I (and others relying on imuxsock for reading from the system log socket) will need to take action to keep rsyslog working as intended on CentOS 7 / RHEL 7 systems. At present (without needed further testing), I believe this will require creation of a systemd unit file drop-in to restore previous behavior.

I'm working on a longer write-up that I will post when I can.

**Comment by ghubusr01** [comment]
@deoren Ok thanks, think I will still wait for that then.

**Comment by deoren** [comment]
I can't speak to how packages worked in the past, but the most recent batch of stable packages had some issues that have been referenced on this ticket.

One package included a rsyslogd binary that I believe was entirely without systemd support. A recent package (`8.31.0-4` ?) fixed that, but still had issues replicating the effects of `systemctl disable rsyslog; systemctl enable rsyslog`. I'm not sure if prior packages handled this differently as I don't recall offhand whether I used the latest upstream packages with CentOS 6 and 7. I don't think I did.

I'm rambling a bit (will try to clean this up before I post it), but I have a few concerns I hope to work through with some additional testing in the next day or so. I'm posting my scratch notes now in case they're useful for further discussion until I can do additional testing:

- How does the `8.31.0-6` package (testing repo) deal with an existing `8.31.0-4` systemd configuration where `systemctl disable rsyslog; systemctl enable rsyslog` has been used to enable the creation of the `/etc/systemd/system/syslog.service` symbolic link (which in turn creates the `/run/systemd/journal/syslog` log socket)? Said another way, is the `/run/systemd/journal/syslog` log socket created on startup now? My configuration relies on imuxsock to read from that socket for syslog messages. The new configuration (both the `/etc/rsyslog.conf` file and the `/usr/lib/systemd/system/rsyslog.service` file) does not depend on either (imjournal module now).

- The `/etc/rsyslog.d/listen.conf` file is provided by the stock CentOS 7 systemd package (need package name here). While I didn't have to comment it out, I did so in my testing. I need to check how this file is handled by the test package.

- I suspect I will need to create `/etc/systemd/system/rsyslog.service.d/descriptive-filename-here.conf` with contents like this to make sure I have the expected `/run/systemd/journal/syslog` log socket created by systemd:

```
[Unit]
Requires=syslog.socket

[Install]
Alias=syslog.service
```

All other settings would be inherited from the `/usr/lib/systemd/system/rsyslog.service` file provided by the `8.31.0-6` package.

More to come once I have a chance to do some testing.

**Comment by deoren** [comment]
I've done some testing and here is what I found.

## Stock CentOS 7 8.24.0 package to 8.31.0-4 stable

Everything said before regarding the workarounds (e.g., `systemctl disable rsyslog; systemctl enable rsyslog`) still applies. I suspect nobody is surprised.


## Stock CentOS 7 8.24.0 package to 8.31.0-10 test package upgrade

The upgrade completes and the same functionality present before is present here. Because the syntax was obsolete legacy format before and the format is obsolete legacy format now the `/etc/rsyslog.d/listen.conf` file passes validation checks (`rsyslogd -N6`) without issue.

That said, the `/etc/rsyslog.d/listen.conf` file doesn't really *do* anything because the `/etc/rsyslog.conf` file disables local logging *and* the `/usr/lib/systemd/system/rsyslog.repo` unit file doesn't enable socket activation (assuming I'm using the right words here, basically the symlink from `/etc/systemd/system/syslog.service` to `/usr/lib/systemd/system/rsyslog.service` wasn't created and systemd doesn't create the `/run/systemd/journal/syslog` socket for rsyslog to read from).

Not a problem here because the conf file was stock before and is still stock (now upstream Adiscon copy), so imjournal is used to pull log messages (API?) instead of via a socket.


## Adiscon repo 8.31.0-4 stable package (with unmodified Adiscon RPM config) to 8.31.0-10 test package upgrade

After installing the 8.31.0-4 package, I ran `systemctl disable rsyslog; systemctl enable rsyslog` and that workaround seemed to allow that version to function as expected (restart, start, stop). I then performed the upgrade to the test package and rebooted. Prior to that, attempting to run `systemctl status rsyslog` warned me that I should run `systemctl daemon-reload` (or restart) to sort things out.

After a restart, all stock settings appeared to function normally. The upgrade (`yum install rsyslog-8.31.0`) pulled in the needed libfastjson package version without my explicitly specifying to install that package. The `/etc/rsyslog.conf` file included in the previous stable version was replaced, but this was to be expected because I did not modify the previous conf file (thus the checksums match).

## Adiscon repo 8.31.0-4 with custom config to 8.31.0-10 test package upgrade

In short, the symlink from `/etc/systemd/system/syslog.service` to `/usr/lib/systemd/system/rsyslog.service` wasn't created and systemd doesn't create the `/run/systemd/journal/syslog` socket for rsyslog to read from. In my setup (I use imuxsock, not imjournal) this means that rsyslog was not able to read from the socket. To restore this functionality, I had to create a drop-in to restore the socket activation (again, I assume I am using the right words here). This file contains the same entries I noted in my previous comment.

Once I did that and either rebooted or ran `systemctl daemon-reload`, the `/run/systemd/journal/syslog` socket was restored.

## Addendum

In my situation I already have two drop-ins (one to cause rsyslog to wait on `network.target` to be reached to reliably forward messages, another to have systemd wait additional time before forcefully shutting down rsyslog), so creating another (or combining them into one perhaps) isn't a real problem. I think it may catch some others who have been using the RPMs off guard however if they're not prepared for the change.

If this is communicated well in the Changelog (RPM and project), announced on the mailing list, forums and perhaps elsewhere, it should help minimize the friction from the upgrade.

Unmodified configurations should continue to work as before, so there is that.

I hope this is useful. Please let me know if you'd like more detail. I would have liked to have gone into more detail now with this post (conf file fragments for illustration), but I don't have that time at the moment. I should have more time later.

**Comment by deoren** [comment]
> @ghubusr01: @deoren Ok thanks, think I will still wait for that then. And forget about this $SystemLogSocketName in /etc/rsyslog.d, this should be empty as far as I can remember on CentOS 7

If you use the advanced format (previously known as "RainerScript"), then you probably had a warning in your `/var/log/messages` file or from validation check (`rsyslogd -N6`) output like this:

```
rsyslogd: command 'SystemLogSocketName' is currently not permitted - did you already set it via a RainerScript command (v6+ config)? [v8.31.0 try http://www.rsyslog.com/e/2222 ]
```

Wiping the contents of the `/etc/rsyslog.d/listen.conf` file or commenting out the offending directive was sufficient to mute the warning.

Thinking back, perhaps it was the 8.30.0 package that had a unit file which was configured similarly to this:

```
[Unit]
...
;Requires=syslog.socket
...

[Install]
...
;Alias=syslog.service
...
```

Without looking back at my notes, I just don't recall. Either way, the `/run/systemd/journal/syslog` socket wasn't available and it was not immediately obvious what the problem was (at least not to me as I didn't really understand whether rsyslog, systemd or some combination of the two was supposed to create the socket).

Like your configuration here:

```
# /etc/rsyslog.conf
module(load="imuxsock" SysSock.Name="/run/systemd/journal/syslog")
module(load="imklog")
input(type="imuxsock" Socket="/dev/log")
```

I ended up specifying `SysSock.Name` so that it referenced that socket. I even shared my findings with others since they appeared to do what was needed. It shouldn't be necessary however, and likely only worked out because a number of users installed a package without systemd support compiled in (which was an accident). By specifying a path to a custom socket that systemd is supposed to be handling, we worked around the issue instead of really solving it.

Instead of the configuration fragment above, this imuxsock load statement should work for 8.31.0-4 (`systemctl disable rsyslog; systemctl enable rsyslog` or equivalent still needed):

```
# /etc/rsyslog.conf

...
module(load="imuxsock")
module(load="imklog")
...
```


```
# /etc/rsyslog.d/listen.conf
# $SystemLogSocketName /run/systemd/journal/syslog
```

For the `/etc/rsyslog.conf` file, note the lack of an input statement that references `/dev/log`. It should not be used for systems that have systemd managing the socket. @rgerhards helped a lot with clarifying the details of how rsyslog, the imuxsock module and systemd interact in regards to the `/dev/log` and `/run/systemd/journal/syslog` log sockets. If you haven't looked at the imuxsock module doc in a while, take a look at the current master version as it includes a lot of additional information vs the current v8-stable version of the doc:

- http://www.rsyslog.com/doc/master/configuration/modules/imuxsock.html
- http://www.rsyslog.com/doc/v8-stable/configuration/modules/imuxsock.html

With the 8.32.0 release the master version of the docs will rollover, so the two will soon be in sync.

If you include a systemd unit drop-in like the following, your configuration should work with the upcoming 8.32.0 CentOS 7 package (assuming that package mirrors the 8.31.0-10 test package settings):

```
# /etc/systemd/system/rsyslog.service.d/custom-settings.conf

[Unit]
Requires=syslog.socket

[Install]
Alias=syslog.service
```

On a bit of a tangent, I recommend adding to that drop-in conf file an additional statement as it helps a lot with systems (in our case virtual machines) that boot up and load rsyslog before the network is fully functional. Without it, rsyslog never seems to break out of whatever loop it is in until we restart it, and *then* remote messages are forwarded (omfwd, omrelp, doesn't seem to matter).

```diff
# /etc/systemd/system/rsyslog.service.d/custom-settings.conf

[Unit]
Requires=syslog.socket

+# rsyslog doesn't seem to gracefully handle lack of functioning name
+# resolution at startup, so here we ask systemd to only
+# start rsyslog once the network target has been reached.
+# This appears to be a bug and has been reported and will likely
+# be fixed at a future date. Until then, this workaround seems to
+# do the trick.
+After=network.target

[Install]
Alias=syslog.service
```

I hope that helps.

**Comment by deoren** [comment]
> @friedl: @deoren Thank you for your feedback. As Rainer said, the changes were intended.
> 
> I updated the RPM a few more times today and fixed the dependency issue for libfastjson by forcing the requirement. In my lab with a fresh install, this seems to work fine now. Can you please confirm?

I covered this in a longer post, but I wanted to respond directly in case my earlier response was lost in the wall of text. ;)

Yes, the dependency does seem to be enforced and the libfastjson package is upgraded when I install just the test package version (`yum install rsyslog-8.31.0`).

**Comment by deoren** [comment]
> @rgerhards: unrelated side-note: we are looking for several years now for folks who would like to help with the packaging, as our best effort for sure is just that - best effort from folks who do not know precisely what they are doing. 

If you're short-handed on staff/volunteers who have packaging expertise, perhaps have someone write-up exactly what it is that you're looking for and post/share it prominently (e.g., perhaps you're covered well on Ubuntu, but need Debian and CentOS depth). Once you have something written up, we could share it around with other lists we're on, retweet your post asking for help on Twitter, etc.

I hope to learn this year what I'll need to start building packages for some minor scripts/etc that I maintain, but probably nothing at the level you guys are working at. If I *do* learn something I think could be useful I'll chime in, but for the time being I'm happy to test and provide feedback when I can. Our shop is a heavy user of Ubuntu, but we also run some CentOS and RHEL boxes and I could perform some testing there also.

> @rgerhards: I was more often then not tempted to stop providing packages...

It's a selfish request I know, but please continue. Even with rough spots in reliability I'd much rather have access to current versions of rsyslog than be stuck on the much older versions that LTS-based distros provide. If it wasn't for your packages I'd probably not have invested the effort into learning about rsyslog that I have. It's not an immediate payoff, but perhaps the long-term benefits are that your community of users is larger because of the investment in providing the packages.

**Comment by ghubusr01** [comment]
@deoren Thanks for your detailed write-up and thoughts, I will wait definitely for 8.32 now and switch from 8.30 to 8.32 then and will first try it on a relatively unimportant production virtuozzo container and try to get everything up there and then "transfer" it to my other servers.

**Comment by davidelang** [comment]
On Sat, 6 Jan 2018, Deoren Moor wrote:

> If you're short-handed on staff/volunteers who have packaging expertise, 
> perhaps have someone write-up exactly what it is that you're looking for and 
> post/share it prominently (e.g., perhaps you're covered well on Ubuntu, but 
> need Debian and CentOS depth). Once you have something written up, we could 
> share it around with other lists we're on, retweet your post asking for help 
> on Twitter, etc.

It's very simple.

The Rsyslog team does not know enough to create correct packages for the 
different distros.

The various distros have maintainers who package rsyslog for their distro, but 
almost none of them help configure the packages provided by rsyslog (and only a 
couple of them actually communicate directly with us, most just ork internally 
to their distro).

Due to the fact that Rsyslog releases happen far more frequently than distro 
releases (Rsyslog releases every 6 weeks), there is a significant demand for 
distro specific packages for the various Rsyslog packages.

We need people ho understand the packaging systems of the different distros, and 
the nuances and policies of the different releases of the distros to help the 
Rsyslog project create packages that will not cause problems if they are 
installed directly over the distro supplied packages.

We currently have scripts that produce packages for Debian, Ubuntu, and Centos 
(on github under https://github.com/rsyslog ), and would welcome any other 
distros.

**Comment by deoren** [comment]
@davidelang Just wanted to respond and acknowledge your reply. I understand all of your points and if there is something I can do to assist please let me know. I won't presume to say what needs to happen, but if someone on the team takes the draft response you've provided and puts it in final form somewhere on the site I'll be happy to assist with spreading the word.

**Comment by rgerhards** [comment]
@deoren I'll see that I put up @davidelang's wording to a decent place.

**Comment by rgerhards** [comment]
I keep this issue open, so that we can track any problems that happen with 8.33 (damn!) here -- thus moving milestone.

**Comment by karloluiten** [comment]
I don't usually do this ("_THIS IS WHAT GITHUB  REACTIONS ARE FOR!!_"), but thanks to all the participants in this  thread for hard work, investigating, testing, fixing stuff and fixing again.

Cheers! Love the product.

**Comment by deoren** [comment]
All,

I upgraded to the `rsyslog-8.32.0-2.el7.centos.x86_64` package and thankfully the behavior matches what I last reported. This made moving back to the former configuration easy. 

For anyone that has been following this thread and has not seen them, here are two news postings that are relevant:

- http://www.rsyslog.com/major-centos7-rpm-changes/
- http://www.rsyslog.com/rsyslog-8-32-0-v8-stable-released/

## Summary

To work around the package switching from using the `imuxsock` module to using the `imjournal` module, I created a systemd drop-in to enable the `/run/systemd/journal/syslog` log socket to be created, rebooted (though I didn't necessarily have to, I just wanted to make sure I knew what the behavior would be from a fresh start) and confirmed that rsyslog was running and forwarding messages to our central log server as expected.

In case it is helpful, I've uploaded to my "rsyslog-examples" repo (it is pretty bare) all but the rsyslog conf fragment that enables message forwarding. Here are the related files:

## rsyslog conf files

- [/etc/rsyslog-centos.conf](https://github.com/deoren/rsyslog-examples/blob/master/etc/rsyslog-centos.conf)
  - rename to `/etc/rsyslogconf` to use
- [/etc/rsyslog.d/listen.conf](https://github.com/deoren/rsyslog-examples/blob/master/etc/rsyslog.d/listen.conf)
  - a very verbose alternative to commenting out the single entry in the original file

## systemd drop-ins

- [/etc/systemd/system/rsyslog.service.d/10-wait-on-network.conf](https://github.com/deoren/rsyslog-examples/blob/master/etc/systemd/system/rsyslog.service.d/10-wait-on-network.conf)
- [/etc/systemd/system/rsyslog.service.d/15-enable-syslog-socket.conf](https://github.com/deoren/rsyslog-examples/blob/master/etc/systemd/system/rsyslog.service.d/15-enable-syslog-socket.conf)
- [/etc/systemd/system/rsyslog.service.d/20-increase-rsyslog-wait-time.conf](https://github.com/deoren/rsyslog-examples/blob/master/etc/systemd/system/rsyslog.service.d/20-increase-rsyslog-wait-time.conf)

Only the `/etc/systemd/system/rsyslog.service.d/15-enable-syslog-socket.conf` file is truly relevant here, but since I find the others useful (especially the `10-wait-on-network.conf` file), I included them as well.

**Comment by rgerhards** [comment]
@deoren I think we can close this issue now - what do you think?

It would be good to save your excellent notes for all to see. Do you think this makes sense inside the doc or on the web site - more or less as a release note. I am undecided.

**Comment by deoren** [comment]
> @rgerhards: @deoren I think we can close this issue now - what do you think?

I think so too. At this point most systems with auto-updates enabled will have already updated and anyone who installs the update manually will have had a chance to do so. You *could* leave it open for another week just for anyone who has yet to upgrade (perhaps a scheduled update/maintenance cycle), but with your sharing the change via the mailing list and on the main website (news section), you have covered the bases.

> @rgerhards: It would be good to save your excellent notes for all to see. Do you think this makes sense inside the doc or on the web site - more or less as a release note. I am undecided.

Not sure. The docs do not seem to have coverage for any specific distro over another, but when working on rsyslog/rsyslog-doc#503 we will have an opportunity to cover this topic.

If on the website (and not the docs generated from the doc project), where would you cover the topic?

**Comment by rgerhards** [comment]
We amended a direct link to the relevant postings here to the release announcement at http://www.rsyslog.com/rsyslog-8-32-0-v8-stable-released/. I consider this to be sufficient, so finally closing this issue.

**Thanks to all who helped solving it!** As a side-note, we got some good new info which we will make use of.

**Comment by wbotelhos** [comment]
Hi @rgerhards ,

I'm having this problem with version **8.34.0**:

```
Job for rsyslog.service failed because a timeout was exceeded. See "systemctl status rsyslog.service" and "journalctl -xe" for details.
```

My `rsyslog -d` mention about PID like you said in the begin:

```
8412.045403755:main thread    : rsyslogd.c: rsyslogd: checking if startup is ok, pidfile '/var/run/rsyslogd.pid'.
rsyslogd: pidfile '/var/run/rsyslogd.pid' and pid 30579 already exist.
If you want to run multiple instances of rsyslog, you need to specify
different pid files for them (-i option).
rsyslogd: run failed with error -3000 (see rsyslog.h or try http://www.rsyslog.com/e/3000 to learn what that number means)
```

- This problem just happen on a fresh machine, since I'm trying create a AMI with [Packer](https://www.packer.io/intro/index.html);
- When a run Packer in debug mode I can log in the temporary machine and try restart manually, but it still fails with timeout;
- Running my [Cookbook](https://github.com/wbotelhos/chef-rsyslog) directly on production (Ubuntu EC2), it works fine;

Logs: [systemctl](https://gist.github.com/wbotelhos/62799cf162ad513e9038753b9a80ad66), [journalctl-xe](https://gist.github.com/wbotelhos/1f551f31c718206a3df9d4a4d187bafb), [rsyslog.service](https://gist.github.com/wbotelhos/4fe92b90f9cb322123761c8f44dfe7e1), [rsyslog-d](https://gist.github.com/wbotelhos/e5d2b1a8814fd463afaf49d107a1a228), []()

**Comment by davidelang** [comment]
There was a request by distro maintainers to have rsyslog no longer create a pid 
file by default when run under systemd. That has created a few problems (mostly 
related to log rotation)

But this sounds like the systemd config for rsyslog needs to be updated to not 
look for the pid file (or pass the -i option to have it create a pid file like 
it used to)

**Comment by wbotelhos** [comment]
@davidelang Do you have a solution for this problem?

**Comment by Mrten** [comment]
I'm seeing this also on Ubuntu 16.04 with rsyslog 8.35.0-0adiscon3trusty1

**Comment by ZedYeung** [comment]
for me, in Ubuntu 18.04, at first, run `systemctl enable rsyslog.service` (I guess the key point is the symlink)
```
Synchronizing state of rsyslog.service with SysV service script with /lib/systemd/systemd-sysv-install.
Executing: /lib/systemd/systemd-sysv-install enable rsyslog
Created symlink /etc/systemd/system/syslog.service → /lib/systemd/system/rsyslog.service.
```
Now, it can sucessfully start with `systemctl start rsyslog`

**Comment by wbotelhos** [comment]
Hi @ZedYeung ,

Here this link is already setted:

```
$ ln -nfs /etc/systemd/system/syslog.service /lib/systemd/system/rsyslog.service
ln: '/etc/systemd/system/syslog.service' and '/lib/systemd/system/rsyslog.service' are the same file
```

The link:

```
lrwxrwxrwx 1 root root 35 Jun 27 08:26 /etc/systemd/system/syslog.service -> /lib/systemd/system/rsyslog.service
```

The content is the original:

```
[Unit]
Description=System Logging Service
Requires=syslog.socket
Documentation=man:rsyslogd(8)
Documentation=http://www.rsyslog.com/doc/

[Service]
Type=notify
ExecStart=/usr/sbin/rsyslogd -n -iNONE
StandardOutput=null
Restart=on-failure

# Increase the default a bit in order to allow many simultaneous
# files to be monitored, we might need a lot of fds.
LimitNOFILE=16384

[Install]
WantedBy=multi-user.target
Alias=syslog.service
```

But I still see:

```
Job for rsyslog.service failed because a timeout was exceeded. See "systemctl status rsyslog.service" and "journalctl -xe" for details.
```

With each version that comes out I try to install, but it does not work. I hope one day to use Rsyslog again.

**Comment by rgerhards** [comment]
@davidelang @jvymazal do you have an idea what the issue is? We do not have RHEL 7 at hand, at CentOS 7 it seems to work (at least as far as I know).

**Comment by jvymazal** [comment]
@rgerhards just tried to quickly skim trough this thread (and was quite surprised how far it span). Regarding RPM-packaging I will gladly help (sadly have next to no experience with ubuntu/debian), however first I would need a re-cap of what is current problem (with Fedora/CentOS/RHEL) because there is, like a year worth of reports and I am not sure which ones still apply.

**Comment by rgerhards** [comment]
@jvymazal would be great if you could lend a helping hand. I guess we can ignore all but the last comments, as these are definitely outdated. Maybe it's time to close this issue and create a new one. Note that I have not seen any issue here and know lots of folks who use the packages with success. But obviously there are some nits.

That said, I would close and lock this issue and create a new one. Does that sound good?

**Comment by jvymazal** [comment]
@rgerhards If the folks who are having problems are ok with it, for me it definitely does :-)

**Comment by wbotelhos** [comment]
Thanks for your help, @jvymazal and of course @rgerhards . :)
